src/as400_sha_fmt_plug.c

```diff
@@ -0,0 +1,419 @@
+	for(i=0;i<(strlen(password));i++)
+	{
+		password_utf16be[(i*2)]=0x00;
+		password_utf16be[(i*2)+1]=password[i];
+	}
+	
+	// Calculate SHA1 hash from userid+password
+	SHA1_Init(&s_ctx);
+
+	SHA1_Update(&s_ctx, userid_utf16be, (AS400_USERID_LENGTH*2));
+	SHA1_Update(&s_ctx, password_utf16be, (strlen(password)*2));
+
+	SHA1_Final(digest, &s_ctx);
+
+	memcpy(hash, digest, sizeof(digest));
+}
+
+
+// ==================================================================================================
+// Main callback to compute as/400 hash. This function is called by JtR engine to
+// calculate a number of AS/400 hashes
+
+static int as400_crypt_all(int *pcount, struct db_salt *salt)
+{	
+	int count = *pcount;
+	int index = 0;
+	
+	// Compute digest for all given plaintext passwords
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index++)
+	{
+		
+		memset(as400_computed_binary_hash[index], 0, BINARY_LENGTH);
+		memset(as400_retrieved_binary_hash[index], 0, BINARY_LENGTH);
+		
+        as400_password_hash(cur_salt->username_utf16be, as400_saved_passwords[index], as400_computed_binary_hash[index]);
+		
+		// Store the binary hash as retrieved from the file (to be used by compare functions)
+		memcpy(as400_retrieved_binary_hash[index], cur_salt->storedbinaryhash, BINARY_LENGTH);
+		
+	}
+	return count;
+}
+
+
+// ==================================================================================================
+// Check if at least one of the computed hashes matches with its associated retrieved hash
+// If this is the case, JtR engine will call cmp_one() to find the matching hashes
+
+static int as400_cmp_all(void *binary,int count)
+{
+	int i;
+	
+	for(i = 0; i < count; i++)
+	{
+		if(!memcmp(as400_computed_binary_hash[i],as400_retrieved_binary_hash[i],BINARY_LENGTH))
+			return 1;
+	}
+
+	return 0;
+}
+
+
+// ==================================================================================================
+// Check which of the calculated hashes matched
+
+static int as400_cmp_one(void *binary,int index)
+{
+	return !memcmp(as400_computed_binary_hash[index],as400_retrieved_binary_hash[index],BINARY_LENGTH);
+}
+
+
+// ==================================================================================================
+// No ASCII ciphertext, thus returns true
+
+static int as400_cmp_exact(char *source,int index)
+{
+	return 1;
+}
+
+
+// ==================================================================================================
+// Structure holding data for internal self tests
+
+static struct fmt_tests as400_tests[] =
+{
+	// Format of each line: {"hash like in internal representation", "valid password for hash", {"userid as in file","hash as in file"}}
+	{"ROB       4C106E52CA196986E1C52C7FCD02AF046B76C73C", "banaan", {"ROB","4C106E52CA196986E1C52C7FCD02AF046B76C73C"}},
+	{"BART      CED8050C275A5005D101051FF5BCCADF693E8AB7", "Kulach007", {"BART","CED8050C275A5005D101051FF5BCCADF693E8AB7"}},
```

Comment:
For john, I think we need a format specific prefix added.

Are the user names always upper case?
Are non-ascii characters allowed? If yes, a sample hash would be good. If not, valid() might be more restrictive.